### PR TITLE
NAS-116683 / 22.12 / Add new helper function to convert mountinfo to dict

### PR DESCRIPTION
**Pull request by anodos325**
We were using libzfs to convert paths to dataset name. Internally
libzfs does this thorugh stat() on the path and looking up
mount entries. We can instead do this ourselves without having
to use a libzfs handle.

To facilitate this add a new general-purpose method to return
a dict containing mount information indexed by devid. The
reason to index this way is because it allows efficient lookup
of an entry based on stat / lstat output. Refactor
filesystem.statvfs to use this as well.

### CODE REVIEW

**Comment by anodos325** [comment]
Will convert filesystem.stat() plugin to use statx in future PR. statx gives some nice benefits like returning btime.

File: src/middlewared/middlewared/plugins/zfs.py [hunk 1]
```diff
@@ -630,12 +632,14 @@ def common_encryption_checks(self, ds):
...
+        st = os.lstat(path)
+        if stat.S_ISLNK(st.st_mode):
+            raise CallError(f'{path}: path is a symbolic link')
+
+        mntinfo = getmntinfo(st.st_dev)[st.st_dev]
+        ds_name = mntinfo['mount_source']
+        if mntinfo['fs_type'] != 'ZFS':
+            raise CallError(f'{path}: not a ZFS filesystem')
 
         if ds_name is not None:
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/zfs.py:644]: Is this possible ?

**Review by sonicaj — COMMENTED** [review]
LGTM, can we add some integration tests and unit tests as well please ? Thanks!

**Comment by anodos325** [comment]
> LGTM, can we add some integration tests and unit tests as well please ? Thanks!

We have coverage via filesystem.statfs tests (which wrap around this). But I can add more explicit tests for path_to_dataset.

**Comment by anodos325** [comment]
f_fsid is unsigned long and so just presenting as string here.

File: src/middlewared/middlewared/plugins/zfs.py [hunk 2]
```diff
@@ -627,19 +628,27 @@ def common_encryption_checks(self, ds):
         if not ds.encrypted:
             raise CallError(f'{id} is not encrypted')
 
-    def path_to_dataset(self, path):
+    def path_to_dataset(self, path, raise_=False):
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/zfs.py:631]: We are not using `raise_` anywhere, let's please remove it and it's usages

File: src/middlewared/middlewared/pytest/unit/utils/test_mountinfo.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    assert mntent['device_id'] == {'major': 0, 'minor': 77, 'dev_t': 77}
+    assert mntent['root'] == '/'
+    assert mntent['mountpoint'] == '/mnt/tank space /Dataset With a space'
+    assert mntent['mount_opts'] == ['RW', 'NOATIME']
+    assert mntent['fs_type'] == 'zfs'
+    assert mntent['mount_source'] == 'tank space /Dataset With a space'
+    assert mntent['super_opts'] == ['RW', 'XATTR', 'POSIXACL']
+
+
+def test__getmntinfo():
```
Comment (sonicaj) [inline src/middlewared/middlewared/pytest/unit/utils/test_mountinfo.py:86]: Would be nice if we tested the actual output of the contents as well to ensure that they actually match to what we are expecting and not just that we were able to parse the fake file.

**Review by sonicaj — APPROVED** [review]

**Pull request merged**